packages/monorepo-scripts/src/prepublish_checks.ts

```diff
@@ -1,9 +1,129 @@
+        const currentVersion = lernaPackage.package.version;
+        const packageName = lernaPackage.package.name;
+        const packageLocation = lernaPackage.location;
+        const nextVersion = await utils.getNextPackageVersionAsync(currentVersion, packageName, packageLocation);
+
+        const localTagVersions = localTagVersionsByPackageName[packageName];
+        if (_.includes(localTagVersions, nextVersion)) {
+            const tagName = `${packageName}@${nextVersion}`;
+            await utils.removeLocalTagAsync(tagName);
+        }
+
+        const remoteTagVersions = remoteTagVersionsByPackageName[packageName];
+        if (_.includes(remoteTagVersions, nextVersion)) {
+            const tagName = `:refs/tags/${packageName}@${nextVersion}`;
+            await utils.removeRemoteTagAsync(tagName);
+        }
+    }
+}
+
+async function checkCurrentVersionMatchesLatestPublishedNPMPackageAsync(
+    updatedPublicLernaPackages: LernaPackage[],
+): Promise<void> {
+    utils.log('Check package versions against npmjs.org...');
+    const versionMismatches = [];
+    for (const lernaPackage of updatedPublicLernaPackages) {
+        const packageName = lernaPackage.package.name;
+        const packageVersion = lernaPackage.package.version;
+        const packageRegistryJsonIfExists = await npmUtils.getPackageRegistryJsonIfExistsAsync(packageName);
+        if (_.isUndefined(packageRegistryJsonIfExists)) {
+            continue; // noop for packages not yet published to NPM
+        }
+        const allVersionsIncludingUnpublished = npmUtils.getPreviouslyPublishedVersions(packageRegistryJsonIfExists);
+        const latestNPMVersion = semverUtils.getLatestVersion(allVersionsIncludingUnpublished);
+        if (packageVersion !== latestNPMVersion) {
+            versionMismatches.push({
+                packageJsonVersion: packageVersion,
+                npmVersion: latestNPMVersion,
+                packageName,
+            });
+        }
+    }
+    if (!_.isEmpty(versionMismatches)) {
+        utils.log(`Found version mismatches between package.json and NPM published versions (might be unpublished).`);
+        _.each(versionMismatches, versionMismatch => {
+            utils.log(
+                `${versionMismatch.packageName}: ${versionMismatch.packageJsonVersion} package.json, ${
+                    versionMismatch.npmVersion
+                } on NPM`,
+            );
+        });
+        throw new Error(`Please fix the above package.json/NPM inconsistencies.`);
+    }
+}
+
+async function checkChangelogFormatAsync(updatedPublicLernaPackages: LernaPackage[]): Promise<void> {
+    utils.log('Check CHANGELOGs for inconsistencies...');
+    const changeLogInconsistencies = [];
+    for (const lernaPackage of updatedPublicLernaPackages) {
+        const packageName = lernaPackage.package.name;
+        const changelog = changelogUtils.getChangelogOrCreateIfMissing(packageName, lernaPackage.location);
```

Comment:
Should we create a changelog in prepublish checks? Maybe just skip if we don't have one?

Reply:
It just creates a file with an empty `[]`. I think that is fine, since it will need to exist eventually.